File: src/main/java/org/jcodec/codecs/h264/io/CAVLC.java
package org.jcodec.codecs.h264.io;

import static org.jcodec.codecs.h264.decode.CAVLCReader.readU;
import static org.jcodec.codecs.h264.decode.CAVLCReader.readZeroBitCount;
import static org.jcodec.common.model.ColorSpace.YUV422;
import static org.jcodec.common.model.ColorSpace.YUV444;

import org.jcodec.codecs.h264.H264Const;
import org.jcodec.codecs.h264.io.model.MBType;
import org.jcodec.codecs.h264.io.model.PictureParameterSet;
import org.jcodec.codecs.h264.io.model.SeqParameterSet;
import org.jcodec.common.io.BitReader;
import org.jcodec.common.io.BitWriter;
import org.jcodec.common.io.VLC;
import org.jcodec.common.model.ColorSpace;
import org.jcodec.common.tools.MathUtil;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Non-CABAC H.264 symbols read/write routines
 * 
 * @author Jay Codec
 * 
 */
public class CAVLC {

    private ColorSpace color;
    private VLC chromaDCVLC;

    private int[] tokensLeft;
    private int[] tokensTop;
    private int mbWidth;
    private int mbMask;

    public CAVLC(SeqParameterSet sps, PictureParameterSet pps, int mbW, int mbH) {
        this.color = sps.chroma_format_idc;
        this.chromaDCVLC = codeTableChromaDC();
        this.mbWidth = sps.pic_width_in_mbs_minus1 + 1;

        this.mbMask = (1 << mbH) - 1;

        tokensLeft = new int[4];
        tokensTop = new int[mbWidth << mbW];
    }

    public int writeACBlock(BitWriter out, int blkIndX, int blkIndY, MBType leftMBType, MBType topMBType, int[] coeff,
            VLC[] totalZerosTab, int firstCoeff, int maxCoeff, int[] scan) {
        VLC coeffTokenTab = getCoeffTokenVLCForLuma(blkIndX != 0, leftMBType, tokensLeft[blkIndY & mbMask],
                blkIndY != 0, topMBType, tokensTop[blkIndX]);

        int coeffToken = writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, coeffTokenTab);

        tokensLeft[blkIndY & mbMask] = coeffToken;
        tokensTop[blkIndX] = coeffToken;

        return coeffToken;
    }

    public void writeChrDCBlock(BitWriter out, int[] coeff, VLC[] totalZerosTab, int firstCoeff, int maxCoeff,
            int[] scan) {
        writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, getCoeffTokenVLCForChromaDC());
    }

    public void writeLumaDCBlock(BitWriter out, int blkIndX, int blkIndY, MBType leftMBType, MBType topMBType,
            int[] coeff, VLC[] totalZerosTab, int firstCoeff, int maxCoeff, int[] scan) {
        VLC coeffTokenTab = getCoeffTokenVLCForLuma(blkIndX != 0, leftMBType, tokensLeft[blkIndY & mbMask],
                blkIndY != 0, topMBType, tokensTop[blkIndX]);

        writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, coeffTokenTab);
    }

    private int writeBlockGen(BitWriter out, int[] coeff, VLC[] totalZerosTab, int firstCoeff, int maxCoeff,
            int[] scan, VLC coeffTokenTab) {
        int trailingOnes = 0, totalCoeff = 0, totalZeros = 0;
        int[] runBefore = new int[maxCoeff];
        int[] levels = new int[maxCoeff];
        for (int i = 0; i < maxCoeff; i++) {
            int c = coeff[scan[i + firstCoeff]];
            if (c == 0) {
                runBefore[totalCoeff]++;
                totalZeros++;
            } else {
                levels[totalCoeff++] = c;
            }
        }
        if (totalCoeff < maxCoeff)
            totalZeros -= runBefore[totalCoeff];

        for (trailingOnes = 0; trailingOnes < totalCoeff && trailingOnes < 3
                && Math.abs(levels[totalCoeff - trailingOnes - 1]) == 1; trailingOnes++)
            ;

        int coeffToken = coeffToken(totalCoeff, trailingOnes);

        coeffTokenTab.writeVLC(out, coeffToken);

        if (totalCoeff > 0) {
            writeTrailingOnes(out, levels, totalCoeff, trailingOnes);
            writeLevels(out, levels, totalCoeff, trailingOnes);

            if (totalCoeff < maxCoeff) {
                totalZerosTab[totalCoeff - 1].writeVLC(out, totalZeros);
                writeRuns(out, runBefore, totalCoeff, totalZeros);
            }
        }
        return coeffToken;
    }

    private void writeTrailingOnes(BitWriter out, int[] levels, int totalCoeff, int trailingOne) {
        for (int i = totalCoeff - 1; i >= totalCoeff - trailingOne; i--)
            out.write1Bit(levels[i] >>> 31);
    }

    private void writeLevels(BitWriter out, int[] levels, int totalCoeff, int trailingOnes) {

        int suffixLen = totalCoeff > 10 && trailingOnes < 3 ? 1 : 0;
        for (int i = totalCoeff - trailingOnes - 1; i >= 0; i--) {
            int absLev = unsigned(levels[i]);
            if (i == totalCoeff - trailingOnes - 1 && trailingOnes < 3)
                absLev -= 2;

            int prefix = absLev >> suffixLen;
            if (suffixLen == 0 && prefix < 14 || suffixLen > 0 && prefix < 15) {
                out.writeNBit(1, prefix + 1);
                out.writeNBit(absLev, suffixLen);
            } else if (suffixLen == 0 && absLev < 30) {
                out.writeNBit(1, 15);
                out.writeNBit(absLev - 14, 4);
            } else {
                if (suffixLen == 0)
                    absLev -= 15;
                int len, code;
                for (len = 12; (code = absLev - (len + 3 << suffixLen) - (1 << len) + 4096) >= (1 << len); len++)
                    ;
                out.writeNBit(1, len + 4);
                out.writeNBit(code, len);
            }
            if (suffixLen == 0)
                suffixLen = 1;
            if (MathUtil.abs(levels[i]) > (3 << (suffixLen - 1)) && suffixLen < 6)
                suffixLen++;
        }
    }

    private final int unsigned(int signed) {
        int sign = signed >>> 31;
        int s = signed >> 31;

        return (((signed ^ s) - s) << 1) + sign - 2;
    }

    private void writeRuns(BitWriter out, int[] run, int totalCoeff, int totalZeros) {
        for (int i = totalCoeff - 1; i > 0 && totalZeros > 0; i--) {
            H264Const.run[Math.min(6, totalZeros - 1)].writeVLC(out, run[i]);
            totalZeros -= run[i];
        }
    }

    public VLC getCoeffTokenVLCForLuma(boolean leftAvailable, MBType leftMBType, int leftToken, boolean topAvailable,
            MBType topMBType, int topToken) {

        int nc = codeTableLuma(leftAvailable, leftMBType, leftToken, topAvailable, topMBType, topToken);

        return H264Const.coeffToken[Math.min(nc, 8)];
    }

    public VLC getCoeffTokenVLCForChromaDC() {
        return chromaDCVLC;
    }

    protected int codeTableLuma(boolean leftAvailable, MBType leftMBType, int leftToken, boolean topAvailable,
            MBType topMBType, int topToken) {

        int nA = leftMBType == null ? 0 : totalCoeff(leftToken);
        int nB = topMBType == null ? 0 : totalCoeff(topToken);

        if (leftAvailable && topAvailable)
            return (nA + nB + 1) >> 1;
        else if (leftAvailable)
            return nA;
        else if (topAvailable)
            return nB;
        else
            return 0;
    }

    protected VLC codeTableChromaDC() {
        if (color == ColorSpace.YUV420J) {
            return H264Const.coeffTokenChromaDCY420;
        } else if (color == YUV422) {
            return H264Const.coeffTokenChromaDCY422;
        } else if (color == YUV444) {
            return H264Const.coeffToken[0];
        }
        return null;
    }

    public int readCoeffs(BitReader _in, VLC coeffTokenTab, VLC[] totalZerosTab, int[] coeffLevel, int firstCoeff,
            int nCoeff, int[] zigzag) {
        int coeffToken = coeffTokenTab.readVLC(_in);
        int totalCoeff = totalCoeff(coeffToken);
        int trailingOnes = trailingOnes(coeffToken);
        // System.out.println("Coeff token. Total: " + totalCoeff +
        // ", trailOne: " + trailingOnes);

        // blockType.getMaxCoeffs();
        // if (blockType == BlockType.BLOCK_CHROMA_DC)
        // maxCoeff = 16 / (color.compWidth[1] * color.compHeight[1]);

        if (totalCoeff > 0) {
            int suffixLength = totalCoeff > 10 && trailingOnes < 3 ? 1 : 0;

            int[] level = new int[totalCoeff];
            int i;
            for (i = 0; i < trailingOnes; i++)
                level[i] = 1 - 2 * _in.read1Bit();

            for (; i < totalCoeff; i++) {
                int level_prefix = readZeroBitCount(_in, "");
                int levelSuffixSize = suffixLength;
                if (level_prefix == 14 && suffixLength == 0)
                    levelSuffixSize = 4;
                if (level_prefix >= 15)
                    levelSuffixSize = level_prefix - 3;

                int levelCode = (Min(15, level_prefix) << suffixLength);
                if (levelSuffixSize > 0) {
                    int level_suffix = readU(_in, levelSuffixSize, "RB: level_suffix");
                    levelCode += level_suffix;
                }
                if (level_prefix >= 15 && suffixLength == 0)
                    levelCode += 15;
                if (level_prefix >= 16)
                    levelCode += (1 << (level_prefix - 3)) - 4096;
                if (i == trailingOnes && trailingOnes < 3)
                    levelCode += 2;

                if (levelCode % 2 == 0)
                    level[i] = (levelCode + 2) >> 1;
                else
                    level[i] = (-levelCode - 1) >> 1;

                if (suffixLength == 0)
                    suffixLength = 1;
                if (Abs(level[i]) > (3 << (suffixLength - 1)) && suffixLength < 6)
                    suffixLength++;
            }

            int zerosLeft;
            if (totalCoeff < nCoeff) {
                if (coeffLevel.length == 4) {
                    zerosLeft = H264Const.totalZeros4[totalCoeff - 1].readVLC(_in);
                } else if (coeffLevel.length == 8) {
                    zerosLeft = H264Const.totalZeros8[totalCoeff - 1].readVLC(_in);
                } else {
                    zerosLeft = H264Const.totalZeros16[totalCoeff - 1].readVLC(_in);
                }
            } else
                zerosLeft = 0;

            int[] runs = new int[totalCoeff];
            int r;
            for (r = 0; r < totalCoeff - 1 && zerosLeft > 0; r++) {
                int run = H264Const.run[Math.min(6, zerosLeft - 1)].readVLC(_in);
                zerosLeft -= run;
                runs[r] = run;
            }
            runs[r] = zerosLeft;

            for (int j = totalCoeff - 1, cn = 0; j >= 0 && cn < nCoeff; j--, cn++) {
                cn += runs[j];
                coeffLevel[zigzag[cn + firstCoeff]] = level[j];
            }
        }

        // System.out.print("[");
        // for (int i = 0; i < nCoeff; i++)
        // System.out.print(coeffLevel[i + firstCoeff] + ", ");
        // System.out.println("]");

        return coeffToken;
    }

    private static int Min(int i, int level_prefix) {
        return i < level_prefix ? i : level_prefix;
    }

    private static int Abs(int i) {
        return i < 0 ? -i : i;
    }

    public static final int coeffToken(int totalCoeff, int trailingOnes) {
        return (totalCoeff << 4) | trailingOnes;
    }

    public static final int totalCoeff(int coeffToken) {
        return coeffToken >> 4;
    }

    public static final int trailingOnes(int coeffToken) {
        return coeffToken & 0xf;
    }

    public static int[] NO_ZIGZAG = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 };

    public void readChromaDCBlock(BitReader reader, int[] coeff, boolean leftAvailable, boolean topAvailable) {
        VLC coeffTokenTab = getCoeffTokenVLCForChromaDC();

        readCoeffs(reader, coeffTokenTab, coeff.length == 16 ? H264Const.totalZeros16
                : (coeff.length == 8 ? H264Const.totalZeros8 : H264Const.totalZeros4), coeff, 0, coeff.length,
                NO_ZIGZAG);
    }

    public void readLumaDCBlock(BitReader reader, int[] coeff, int mbX, boolean leftAvailable, MBType leftMbType,
            boolean topAvailable, MBType topMbType, int[] zigzag4x4) {
        VLC coeffTokenTab = getCoeffTokenVLCForLuma(leftAvailable, leftMbType, tokensLeft[0], topAvailable, topMbType,
                tokensTop[mbX << 2]);

        readCoeffs(reader, coeffTokenTab, H264Const.totalZeros16, coeff, 0, 16, zigzag4x4);
    }

    public int readACBlock(BitReader reader, int[] coeff, int blkIndX, int blkIndY, boolean leftAvailable,
            MBType leftMbType, boolean topAvailable, MBType topMbType, int firstCoeff, int nCoeff, int[] zigzag4x4) {
        VLC coeffTokenTab = getCoeffTokenVLCForLuma(leftAvailable, leftMbType, tokensLeft[blkIndY & mbMask],
                topAvailable, topMbType, tokensTop[blkIndX]);

        int readCoeffs = readCoeffs(reader, coeffTokenTab, H264Const.totalZeros16, coeff, firstCoeff, nCoeff, zigzag4x4);
        tokensLeft[blkIndY & mbMask] = tokensTop[blkIndX] = readCoeffs;

        return totalCoeff(readCoeffs);
    }

    public void setZeroCoeff(int blkIndX, int blkIndY) {
        tokensLeft[blkIndY & mbMask] = tokensTop[blkIndX] = 0;
    }
}

File: src/main/java/org/jcodec/codecs/h264/io/CAVLC.kt
package org.jcodec.codecs.h264.io

import org.jcodec.codecs.h264.H264Const
import org.jcodec.codecs.h264.io.model.MBType
import org.jcodec.codecs.h264.io.model.PictureParameterSet
import org.jcodec.codecs.h264.io.model.SeqParameterSet
import org.jcodec.common.io.BitReader
import org.jcodec.common.io.BitWriter
import org.jcodec.common.io.VLC
import org.jcodec.common.model.ColorSpace

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Non-CABAC H.264 symbols read/write routines
 *
 * @author Jay Codec
 */
class CAVLC private constructor(private val color: ColorSpace?, mbWidth: Int, mbW: Int, mbH: Int) {
    val coeffTokenVLCForChromaDC: VLC?
    private val tokensLeft: IntArray
    private val tokensTop: IntArray
    private val mbWidth: Int
    private val mbMask: Int
    private val mbW: Int
    private val mbH: Int

    constructor(sps: SeqParameterSet, pps: PictureParameterSet?, mbW: Int, mbH: Int) : this(sps.chromaFormatIdc, sps.picWidthInMbsMinus1 + 1, mbW, mbH) {}

    fun fork(): CAVLC {
        val ret = CAVLC(color, mbWidth, mbW, mbH)
        System.arraycopy(tokensLeft, 0, ret.tokensLeft, 0, tokensLeft.size)
        System.arraycopy(tokensTop, 0, ret.tokensTop, 0, tokensTop.size)
        return ret
    }

    fun writeACBlock(out: BitWriter?, blkIndX: Int, blkIndY: Int, leftMBType: MBType?, topMBType: MBType?, coeff: IntArray?,
                     totalZerosTab: Array<VLC?>?, firstCoeff: Int, maxCoeff: Int, scan: IntArray?): Int {
        val coeffTokenTab = getCoeffTokenVLCForLuma(blkIndX != 0, leftMBType, tokensLeft[blkIndY and mbMask],
                blkIndY != 0, topMBType, tokensTop[blkIndX])
        val coeffToken = CAVLCUtil.writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, coeffTokenTab)
        tokensLeft[blkIndY and mbMask] = coeffToken
        tokensTop[blkIndX] = coeffToken
        return coeffToken
    }

    fun writeChrDCBlock(out: BitWriter?, coeff: IntArray?, totalZerosTab: Array<VLC?>?, firstCoeff: Int, maxCoeff: Int,
                        scan: IntArray?) {
        CAVLCUtil.writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, coeffTokenVLCForChromaDC)
    }

    fun writeLumaDCBlock(out: BitWriter?, blkIndX: Int, blkIndY: Int, leftMBType: MBType?, topMBType: MBType?,
                         coeff: IntArray?, totalZerosTab: Array<VLC?>?, firstCoeff: Int, maxCoeff: Int, scan: IntArray?) {
        val coeffTokenTab = getCoeffTokenVLCForLuma(blkIndX != 0, leftMBType, tokensLeft[blkIndY and mbMask],
                blkIndY != 0, topMBType, tokensTop[blkIndX])
        CAVLCUtil.writeBlockGen(out, coeff, totalZerosTab, firstCoeff, maxCoeff, scan, coeffTokenTab)
    }

    protected fun codeTableChromaDC(): VLC? {
        if (color == ColorSpace.YUV420J) {
            return H264Const.coeffTokenChromaDCY420
        } else if (color == ColorSpace.YUV422) {
            return H264Const.coeffTokenChromaDCY422
        } else if (color == ColorSpace.YUV444) {
            return H264Const.CoeffToken[0]
        }
        return null
    }

    fun readChromaDCBlock(reader: BitReader?, coeff: IntArray, leftAvailable: Boolean, topAvailable: Boolean) {
        val coeffTokenTab = coeffTokenVLCForChromaDC
        CAVLCUtil.readCoeffs(reader, coeffTokenTab, if (coeff.size == 16) H264Const.totalZeros16 else if (coeff.size == 8) H264Const.totalZeros8 else H264Const.totalZeros4, coeff, 0, coeff.size,
                NO_ZIGZAG)
    }

    fun readLumaDCBlock(reader: BitReader?, coeff: IntArray?, mbX: Int, leftAvailable: Boolean, leftMbType: MBType?,
                        topAvailable: Boolean, topMbType: MBType?, zigzag4x4: IntArray?) {
        val coeffTokenTab = getCoeffTokenVLCForLuma(leftAvailable, leftMbType, tokensLeft[0], topAvailable, topMbType,
                tokensTop[mbX shl 2])
        CAVLCUtil.readCoeffs(reader, coeffTokenTab, H264Const.totalZeros16, coeff, 0, 16, zigzag4x4)
    }

    fun readACBlock(reader: BitReader?, coeff: IntArray?, blkIndX: Int, blkIndY: Int, leftAvailable: Boolean,
                    leftMbType: MBType?, topAvailable: Boolean, topMbType: MBType?, firstCoeff: Int, nCoeff: Int, zigzag4x4: IntArray?): Int {
        val coeffTokenTab = getCoeffTokenVLCForLuma(leftAvailable, leftMbType, tokensLeft[blkIndY and mbMask],
                topAvailable, topMbType, tokensTop[blkIndX])
        val readCoeffs = CAVLCUtil.readCoeffs(reader, coeffTokenTab, H264Const.totalZeros16, coeff, firstCoeff, nCoeff, zigzag4x4)
        tokensTop[blkIndX] = readCoeffs
        tokensLeft[blkIndY and mbMask] = tokensTop[blkIndX]
        return totalCoeff(readCoeffs)
    }

    fun setZeroCoeff(blkIndX: Int, blkIndY: Int) {
        tokensTop[blkIndX] = 0
        tokensLeft[blkIndY and mbMask] = tokensTop[blkIndX]
    }

    init {
        coeffTokenVLCForChromaDC = codeTableChromaDC()
        this.mbWidth = mbWidth
        mbMask = (1 shl mbH) - 1
        this.mbW = mbW
        this.mbH = mbH
        tokensLeft = IntArray(4)
        tokensTop = IntArray(mbWidth shl mbW)
    }

    companion object {

        fun Min(i: Int, level_prefix: Int): Int {
            return if (i < level_prefix) i else level_prefix
        }

        fun Abs(i: Int): Int {
            return if (i < 0) -i else i
        }

        fun totalCoeff(coeffToken: Int): Int {
            return coeffToken shr 4
        }

        fun trailingOnes(coeffToken: Int): Int {
            return coeffToken and 0xf
        }

        fun codeTableLuma(leftAvailable: Boolean, leftMBType: MBType?, leftToken: Int, topAvailable: Boolean,
                          topMBType: MBType?, topToken: Int): Int {
            val nA = if (leftMBType == null) 0 else totalCoeff(leftToken)
            val nB = if (topMBType == null) 0 else totalCoeff(topToken)
            return if (leftAvailable && topAvailable) nA + nB + 1 shr 1 else if (leftAvailable) nA else if (topAvailable) nB else 0
        }

        fun getCoeffTokenVLCForLuma(leftAvailable: Boolean, leftMBType: MBType?, leftToken: Int, topAvailable: Boolean,
                                    topMBType: MBType?, topToken: Int): VLC? {
            val nc = codeTableLuma(leftAvailable, leftMBType, leftToken, topAvailable, topMBType, topToken)
            return H264Const.CoeffToken[Math.min(nc, 8)]
        }

        fun writeTrailingOnes(out: BitWriter, levels: IntArray, totalCoeff: Int, trailingOne: Int) {
            for (i in totalCoeff - 1 downTo totalCoeff - trailingOne) out.write1Bit(levels[i] ushr 31)
        }

        fun unsigned(signed: Int): Int {
            val sign = signed ushr 31
            val s = signed shr 31
            return ((signed xor s) - s shl 1) + sign - 2
        }

        fun writeRuns(out: BitWriter?, run: IntArray, totalCoeff: Int, totalZeros: Int) {
            var totalZeros = totalZeros
            var i = totalCoeff - 1
            while (i > 0 && totalZeros > 0) {
                H264Const.run[Math.min(6, totalZeros - 1)].writeVLC(out, run[i])
                totalZeros -= run[i]
                i--
            }
        }

        val NO_ZIGZAG = intArrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)

    }
}